### ISSUE TITLE
Long-term home for core Portal Network types

### ISSUE THREAD

**Issue by njgheorghita** [ethereum/trin#606]
@ogenev and have been discussing this for the past days, and we think it's time to open it up with the group to get feedback and decide on a path forward.

The core question is where will we define core portal network types that we specifically want to expose over jsonrpc. 
1. Inside `ethportal-api`
2. Inside `trin-types`. 

For example. Let's take `HeaderWithProof`. It's required by `ethportal-api` as well as `trin-history`, `trin-validation` and `trin-bridge` as well as 3rd party libraries using `ethportal-api`.

(Currently this type is defined twice, once inside `ethportal-api` and once inside `trin-types`. but @ogenev and I are in agreement that we should move towards a single definition.)

The question is essentially, which workspace should act as the "root" level package where we store these type definitions.
`ethportal-api` -> `trin-types` -> `trin-...`
or
`trin-types` -> `ethportal-api` -> `trin-...`

We'll leave some thoughts below for everyone to ponder and then we can discuss it tomorrow during sync

**Comment by njgheorghita** [comment]
IMO, we should follow the convention set in the `README` for `trin-types`.
`Library for shared types across trin workspace members.`

If we start defining core types inside `ethportal-api`. Then it will become a de-facto "secondary" `trin-types`, where we will end up defining many of our core types. It seems much more organized to simply define all shared types in one place. 

One other way to frame this question is...
My opinion:  the `trin` in `trin-types` refers to all kinds of types used inside this project directory.
Oggie's opinion: the `trin` in `trin-types` refers to shared types specifically used by the `trin` client. The core types we're debating are better categorized as "portal" types - since they're also used by external users of `ethportal-api` (eg glados).

Tbf there is more nuance in this decision, including versioning dependencies, the `reth` situation & how we plan to resolve that. But I'll leave that for discussion tm.

**Comment by carver** [comment]
Preliminary proposal from the Thursday call, waiting on @ogenev 's thoughts on Monday: define all common types in trin-types. Don't *implement* any types in ethportal-api, only re-export them from their trin-types implementation.

**Comment by perama-v** [comment]
> Preliminary proposal from the Thursday call, waiting on @ogenev 's thoughts on Monday: define all common types in trin-types. Don't _implement_ any types in ethportal-api, only re-export them from their trin-types implementation.

This sounds good. Defining all in one place is good and `glados` is flexible to such a change

**Comment by carver** [comment]
Came up in another PR: should we publish a `trin-types` crate for other projects to use directly, or always publish types through ethportal-api?

**Comment by perama-v** [comment]
Mild preference: Seems simpler to publish `trin-types` crate compared to exposing through `ethportal-api`

**Comment by njgheorghita** [comment]
Imo, it's worthwhile publishing `trin-types` regardless. As to whether we re-export some types in `ethportal-api` - or simply require users of `ethportal-api` to also use `trin-types`.... I'm not sure. It seems to me worthwhile to re-export the most basic types. So the most basic consumer of `ethportal-api` doesn't need to use `trin-types`. But, at some point, defining this "basic user" will become difficult as we extend `ethportal-api` features. So, it might make sense just to enforce all users to depend on both libraries from the get-go. In the same way that we've had to depend on `ethereum-types` ever since we've been using `discv5`
